### ISSUE TITLE
Networking: Support overlay networking with docker swarm

### ISSUE THREAD

**Issue by mcastelino** [01org/cc-oci-runtime#95]
Currently the networking implementation supports network=none and network=bridge. 
Docker swarm overlay networking needs to be supported to enable clusterwide networking.

Requires support for https://github.com/01org/cc-oci-runtime/issues/94

Swarm sets up two veth interfaces per container
- overlay veth
  - with IP Addresses (Container Overlay IP and Service Virtual IP)
- docker_gw_bridge veth
  - with a host level IP address for non overlay traffic

The docker_gw_bridge is used for non overlay traffic (if the overlay network configuration allows for this i.e. --internal=true is not set)
### Discovery

Runtime MUST discover
- Hostname
- All interfaces and their names
- All IP’s on interfaces
- All routes
- macaddress, MTU’s and other properties per interface
### Connectivity

Runtime MUST 
- Connect all the interfaces to QEMU and setup macaddress
- Handle all interface types (veth, ovs, tap …)
### Configuration:

Runtime MUST
- Pass all of this information into the VM

VM cloud-init service MUST
- Setup Hostname
- Setup Interface name, MTU, IPv4 addresses and subnet masks, IPv6 addresses
- Setup routes

**Comment by mcastelino** [comment]
Here is a typical example of a container within a docker swarm

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
40: eth0@if41: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP mode DEFAULT group default
    link/ether 02:42:0a:00:00:0e brd ff:ff:ff:ff:ff:ff
44: eth1@if45: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default
    link/ether 02:42:ac:12:00:07 brd ff:ff:ff:ff:ff:ff

```

Here 
- eth0:  is the veth on the overlay network
- eth1:  is the veth on the docker_gw_bridge (which may or may not be present)

They have the following IP configuration

```

40: eth0@if41: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UP group default
    link/ether 02:42:0a:00:00:0e brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.14/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.0.3/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:aff:fe00:e/64 scope link
       valid_lft forever preferred_lft forever
44: eth1@if45: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:ac:12:00:07 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.7/16 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe12:7/64 scope link
       valid_lft forever preferred_lft forever

```

On eth0 you will notice two IP addresses

- 10.0.0.14: which is the container overlay IP address
- 10.0.0.3: which is the Virtual IP of the "Docker Service" that this container is part of

On eth1  you will notice 172.18.0.7 which is a host local IP used to exit directly out of the container (if allowed) NATed to the main network

The route setup is as follows

```
default via 172.18.0.1 dev eth1
10.0.0.0/24 dev eth0  proto kernel  scope link  src 10.0.0.14
172.18.0.0/16 dev eth1  proto kernel  scope link  src 172.18.0.7

```

Here the default gateway is setup to be eth1, so only the overlay traffic uses overlay network, all other traffic uses the docker_gateway_br network.

Hence to support swarm we need to support
- adding these interfaces into the VM
- ensuring that they have the right name in side the VM
- ensure that multiple IPs can be assigned to the interface (as needed)
- ensure that the routes are setup as needed

**Comment by grahamwhaley** [comment]
Hi @mcastelino I know #167 was a partial fix for some of this item - can you give a brief update here on what is left or reference another ticket if appropriate?  Thanks.

**Comment by jodh-intel** [comment]
This is blocked on hyperstart (#271).

**Comment by amshinde** [comment]
Closing this as docker swarm integration tests have been merged with #524
